nio/client/base_client.py

```diff
@@ -129,11 +129,17 @@ class ClientCallback:
     func: Union[Callable[..., None], Callable[..., Awaitable[None]]] = field()
     filter: Union[Tuple[Type, ...], Type, None] = None
 
-    async def execute(self, event, room: Optional[MatrixRoom] = None) -> None:
+    def execute(self, event, room: Optional[MatrixRoom] = None) -> Optional[Awaitable]:
         if self.filter is None or isinstance(event, self.filter):
-            result = self.func(room, event) if room else self.func(event)
-            if inspect.isawaitable(result):
-                await result
+            if room:
+                return self.func(room, event)
+            else:
+                return self.func(event)
+
+    async def async_execute(self, event, room: Optional[MatrixRoom] = None) -> None:
+        result = self.execute(event, room)
+        if inspect.isawaitable(result):
+            await result
```

Comment:
In practice, both the sync and async clients need to attempt to invoke a callback as either sync or async; someone can reasonably pass a sync (and hopefully, nonblocking) callback to an async client and it *ought* to work

Reply:
There wasn't one method for each before. In stead, in places where it was done solely synchronously, it was hard-coded to loop and do the if-statements. All this does is to put that into a method to avoid repetition. And to further avoid repetition, the asynchronous method just calls the synchronous one; if an awaitable is returned, await it.

Everything should work exactly as before. Are you *suggesting* the synchronous method should have an extra step added for if the function is asynchronous? For example:

```python
def execute(self, event, room: Optional[MatrixRoom] = None) -> Optional[Awaitable]:
    if self.filter is None or isinstance(event, self.filter):
        if room:
            result = self.func(room, event)
        else:
            result =  self.func(event)

        if inspect.isawaitable(result):
            asyncio.run(result)
```

Reply:
_(I noticed I hadn't included the `filter` if-statement in the asynchronous method in my commit. I fixed that just now and force pushed.)_